subsys/usb/usb_c/usbc_pe_common.c

```diff
@@ -720,6 +733,216 @@ void pe_drs_send_swap_exit(void *obj)
 	usbc_timer_stop(&pe->pd_t_sender_response);
 }
 
+/**
+ * @brief 8.3.3.4.2.1 PE_SNK_Send_Soft_Reset State
+ */
+void pe_soft_reset_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/* Reset the protocol layer */
+	prl_reset(dev);
+
+	/* Initialize PE Submachine */
+	pe->submachine = 0;
+}
+
+void pe_soft_reset_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	if (prl_is_running(dev)) {
+		switch (pe->submachine) {
+		case 0:
+			/* Send Accept message to SOP */
+			pe_send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_ACCEPT);
+			/* Start Sender Response Timer */
+			usbc_timer_start(&pe->pd_t_sender_response);
+			/* Move to next substate */
+			pe->submachine++;
+			break;
+		case 1:
+			/*
+			 * The Policy Engine Shall transition to the
+			 * PE_SRC_Send_Capabilities state when:
+			 *      1: Accept message sent to SOP
+			 */
+			if (atomic_test_and_clear_bit(pe->flags, PE_FLAGS_TX_COMPLETE)) {
+				pe_set_state(dev, PE_SNK_WAIT_FOR_CAPABILITIES);
+			}
+			/*
+			 * The Policy Engine Shall transition to the
+			 * PE_SRC_Hard_Reset state when:
+			 *      1: A SenderResponseTimer timeout occurs.
+			 *      2: OR the Protocol Layer indicates that a
+			 *         transmission error has occurred.
+			 *         (Handled in pe_report_error)
+			 */
+			else if (usbc_timer_expired(&pe->pd_t_sender_response)) {
+				pe_set_state(dev, PE_SNK_HARD_RESET);
+			}
+			break;
+		}
```

Comment:
This is an instance where the spec has a single state, but within that state it has a number of things to check off its list before we move onto the next state.  Ideally, we'll code this in a way which is both easy to read code while still easily correlating to the spec. 

It looks like we send the Accept from entry in our own code, which avoids the sub state.  I'm assuming we don't do that here to avoid sending the message when the PRL isn't running (and is that an improvement we should make in our own stack as well)?

Reply:
Since the substates are unique to this state, what if we defined a single-purpose enum for the substates and separate run functions for each substate? Would that resolve readability concerns?

Reply:
I generally agree with you Keith but numbered states seems to be an accepted practice in Zephyr:
https://github.com/zephyrproject-rtos/zephyr/blob/7b5780e0076434a01543928245538300b44a3785/subsys/usb/device/class/dfu/usb_dfu.c#L810-L815
https://github.com/zephyrproject-rtos/zephyr/blob/7b5780e0076434a01543928245538300b44a3785/subsys/emul/emul_bmi160.c#L98-L104
https://github.com/zephyrproject-rtos/zephyr/blob/7b5780e0076434a01543928245538300b44a3785/subsys/net/lib/coap/coap.c#L318-L327

Reply:
Done